adc/ad7779/imx6ull/dma.c

```diff
@@ -0,0 +1,129 @@
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+#include <sys/interrupt.h>
+
+#include <phoenix/arch/imx6ull.h>
+#include <sdma.h>
+
+#include "../ad7779.h"
+
+#define SDMA_DEVICE_FILE_NAME ("/dev/sdma/ch07")
+
+struct {
+	sdma_t sdma;
+	struct {
+		size_t size;
+		size_t count;
+		void *ptr;
+	} buffer;
+	sdma_buffer_desc_t *bd;
+} sdma;
+
+static int sdma_configure(size_t size, size_t count, addr_t *phys_addr)
+{
+	int res;
+
+	uint8_t event_transfer = 39; /* SAI3 RX FIFO */
+	uint8_t event_channel = event_transfer;
+
+	unsigned tries = 25;
+	while ((res = sdma_open(&sdma.sdma, SDMA_DEVICE_FILE_NAME)) < 0) {
+		usleep(100 * 1000);
+		if (--tries == 0) {
+			log_error("failed to open SDMA device file (%s)", SDMA_DEVICE_FILE_NAME);
+			return -1;
+		}
+	}
+
+	sdma.buffer.ptr = sdma_alloc_uncached(&sdma.sdma, size, phys_addr, 1);
+	if (sdma.buffer.ptr == NULL) {
+		log_error("failed to allocate buffers");
+		return -1;
+	}
+
+	memset(sdma.buffer.ptr, 0, size);
+
+	addr_t bd_paddr;
+	sdma.bd = sdma_alloc_uncached(&sdma.sdma, count * sizeof(sdma_buffer_desc_t), &bd_paddr, 1);
+	if (sdma.bd == NULL) {
+		log_error("failed to allocate memory for buffer descriptors");
+		return -1;
+	}
+
+	for (int i = 0; i < count; i++) {
```

Comment:
C89
also: use `unsigned int` when comparing against `size_t` (to avoid implicit type conversion)